File: tests/test_bron.py
def printState(R,P,X):
    print('R:' + str(R))
    print('P:' + str(P))
    print('X:' + str(X))
    print('\n')

def bronk2(cliques, R, P, X, g):
    if not any((P, X)):
        cliques.append(R)
    for v in P[:]:
        R_v = R + [v]
        P_v = [v1 for v1 in P if v1 in g[v]]
        X_v = [v1 for v1 in X if v1 in g[v]]
        bronk2(cliques, R_v, P_v, X_v, g)
        P.remove(v)
        X.append(v)

graph = [[13,37],[16],[8, 11, 20],[35], [6,21,7,17,8,9,13,16,12],

        [], [4,7,8,9,12,16,17,21,10,13], [4,6,13,15,16,12,17,9,11,10], [4,6,2,10,11,12,13,15,17,9,16],

        [4,6,7,8,10,11,12,17,23,21,20],[8,9,6,7,11,12,13], [7,8,9,10,12,13,2], [6,7,8,9,10,11,4,17,21],

        [0,4,7,8,10,11,6,37], [], [7,8,31,35], [1,4,6,7,8,17],

        [4,6,7,8,9,12,16,21,22,23,25], [19,20], [18,20,36], [18,19,9,2,36],

        [4,6,9,12,17,23], [17,27,28,24], [9,17,21,24,28,27,25], [22,23,30,25,26,27,28],

        [17,23,24,27,30], [24,30], [22,23,24,25,28], [22,23,27,24],

        [32,34,33], [24,26,25], [15], [29,34],

        [29], [29,32], [15,3], [19,20],

        [0,13], []]

cliques = []
bronk2(cliques, [], list(range(39)) ,[],graph)
cliques.sort(key=len, reverse=True)
[print(i) for i in cliques if len(i) == 6]

File: tests/test_bron.cpp
#include <iostream>
#include <sstream>
#include <ostream>
#include <cstdio>
#include <cstdlib>
#include <string>
#include <vector>
#include <utility>
#include <algorithm>
#include <set>

#define MAX_VERTEX 39

using namespace std;

void printState(vector<int> R, vector<int> P, vector<int> X)
{
  cout << "BEGIN STATE" << endl;

  cout << "R :" << endl;
  for(int k = 0; k < R.size(); ++k)
    cout << R[k] << " " << endl;

  cout << endl;

  cout << "P :" << endl;
  for(int k = 0; k < P.size(); ++k)
    cout << P[k] << " " << endl;

  cout << endl;

  cout << "X :" << endl;
  for(int k = 0; k < X.size(); ++k)
    cout << X[k] << " " << endl;

  cout << endl;

  cout << "END STATE" << endl;
}

void printClique(vector<int> clique)
{
  for(int k = 0; k < clique.size(); ++k)
    cout << clique[k] << " " << endl;

  cout << " ------ NEXT CLIQUE ------- " << endl;
}

vector<int> uni(vector<int> first, int second)
{
  first.push_back(second);

  return first;
}

vector<int> intersec(vector<vector<int> > graph, vector<int> conf, int second)
{
  vector<int> newfirst;

  for(int k = 0; k < conf.size(); ++k)
  {
    for(int l = 0; l < graph[second].size(); ++l)
    {
      if(conf[k] == graph[second][l])
      {
        newfirst.push_back(conf[k]);
      }

    }
  }

  return newfirst;
}

void bron(vector<vector<int> > graph, vector<int> R,vector<int> P,vector<int> X)
{
  vector<int> newr, newp, newx;

  if(P.empty() && X.empty())
  {
    cout << "R maximal" << endl;
    printClique(R);
  }
  else
  {
    for(int v : P)
    {
      newr = uni(R, v);
      newp = intersec(graph, P, v);
      newx = intersec(graph, X, v);
      bron(graph, newr, newp , newx);
      P.erase(remove(P.begin(), P.end(), v), P.end());
      X.push_back(v);
    }
  }
}

void prepareBron()
{
  vector<int> R, P, X;
  vector<vector<int> > graph(4);

  graph[0].push_back(1);
  graph[0].push_back(2);
  graph[1].push_back(0);
  graph[1].push_back(2);
  graph[1].push_back(3);
  graph[2].push_back(0);
  graph[2].push_back(1);
  graph[3].push_back(1);

  for(int i = 0; i < 4; i++)
    P.push_back(i);

  bron(graph, R, P, X);
}

int main()
{
  prepareBron();

  return 0;
}
